In [79]:
from sklearn import datasets
from sklearn.ensemble import BaggingClassifier as bagging
from sklearn.ensemble import AdaBoostClassifier as adaboost
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import KFold, cross_val_score, train_test_split
from numpy import mean, std
from sklearn.metrics import accuracy_score
iris_data = datasets.load_iris()

In [80]:
X = iris_data.data
y = iris_data.target

In [85]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size =0.25, random_state = 0)

In [92]:
accuracy = []
for c in range(1, 11):
    bagging_model = bagging(DecisionTreeClassifier(max_depth = 1), n_estimators = c, random_state = 0).fit(X_train, y_train)
    bagging_model_predictions = bagging_model.predict(X_test)
    bagging_model_accuracy = accuracy_score(bagging_model_predictions, y_test)
    print("Bagging model score using " +str(c)+" estimators: " +str(bagging_model_accuracy))
    accuracy.append(bagging_model_accuracy)

Bagging model score using 1 estimators: 0.7631578947368421
Bagging model score using 2 estimators: 0.5789473684210527
Bagging model score using 3 estimators: 0.5789473684210527
Bagging model score using 4 estimators: 0.8947368421052632
Bagging model score using 5 estimators: 0.8947368421052632
Bagging model score using 6 estimators: 0.8947368421052632
Bagging model score using 7 estimators: 0.9736842105263158
Bagging model score using 8 estimators: 0.9736842105263158
Bagging model score using 9 estimators: 0.9736842105263158
Bagging model score using 10 estimators: 0.9736842105263158


In [97]:
accuracy1 = []
for d in range(1, 11):
    ada_model = adaboost(DecisionTreeClassifier(max_depth = 1), n_estimators = d, random_state = 0).fit(X_train, y_train)
    ada_model_predictions = ada_model.predict(X_test)
    ada_model_accuracy = accuracy_score(ada_model_predictions, y_test)
    print("ada boost model score using " +str(d)+" estimators: " +str(ada_model_accuracy))
    accuracy1.append(ada_model_accuracy)

ada boost model score using 1 estimators: 0.5789473684210527
ada boost model score using 2 estimators: 0.8947368421052632
ada boost model score using 3 estimators: 0.8947368421052632
ada boost model score using 4 estimators: 0.9736842105263158
ada boost model score using 5 estimators: 0.9736842105263158
ada boost model score using 6 estimators: 0.8947368421052632
ada boost model score using 7 estimators: 0.8947368421052632
ada boost model score using 8 estimators: 0.9736842105263158
ada boost model score using 9 estimators: 0.9736842105263158
ada boost model score using 10 estimators: 0.8947368421052632


The accuracy of the bagging classifier increased intially and was maintained contstant by increasing the number of estimators. Increasing the number of estimators in the boosting classifier caused an early fluctuation in the model's score. However, the final score was found to be a factor of 0.08 lower than when only one estimator was utilized

In [100]:

accuracy_bag = []
for j in range(1, 11):
    bagclassfifier = bagging(DecisionTreeClassifier(max_depth = 1).fit(X_train,y_train), n_estimators= j, random_state = 0) 
    scores = cross_val_score( bagclassfifier, X, y, cv = 5)
    print("n_estmators = %d, 'Accuracy: %.3f, %.3f'" % (j, mean(scores), std(scores)))
    accuracy_bag.append(scores)

n_estmators = 1, 'Accuracy: 0.640, 0.025'
n_estmators = 2, 'Accuracy: 0.860, 0.169'
n_estmators = 3, 'Accuracy: 0.927, 0.053'
n_estmators = 4, 'Accuracy: 0.927, 0.053'
n_estmators = 5, 'Accuracy: 0.927, 0.053'
n_estmators = 6, 'Accuracy: 0.927, 0.053'
n_estmators = 7, 'Accuracy: 0.927, 0.053'
n_estmators = 8, 'Accuracy: 0.927, 0.053'
n_estmators = 9, 'Accuracy: 0.927, 0.053'
n_estmators = 10, 'Accuracy: 0.927, 0.053'


In [96]:
accuracy_ada = []
for i in range(1, 11):
    adaclassfifier = adaboost(DecisionTreeClassifier(max_depth = 1).fit(X_train,y_train), n_estimators= i, random_state = 0) 
    scores = cross_val_score(adaclassfifier, X, y, cv = 5)
    print("n_estmators = %d, 'Accuracy: %.3f, %.3f'" % (i, mean(scores), std(scores)))
    accuracy_ada.append(scores)


n_estmators = 1, 'Accuracy: 0.667, 0.000'
n_estmators = 2, 'Accuracy: 0.947, 0.034'
n_estmators = 3, 'Accuracy: 0.947, 0.034'
n_estmators = 4, 'Accuracy: 0.953, 0.027'
n_estmators = 5, 'Accuracy: 0.953, 0.027'
n_estmators = 6, 'Accuracy: 0.887, 0.115'
n_estmators = 7, 'Accuracy: 0.840, 0.145'
n_estmators = 8, 'Accuracy: 0.900, 0.119'
n_estmators = 9, 'Accuracy: 0.953, 0.027'
n_estmators = 10, 'Accuracy: 0.947, 0.034'


Using cross-validation, increasing the number of estimators in the bagging classifier improved the model's performance from 0.640 to 0.927. In the case of the ada boosting classifier, increasing the number of estimators caused accuracy to fluctuate, with the final accuracy being somewhat lower than when only one estimator was used.